src/goto-programs/read_bin_goto_object.cpp

```diff
@@ -254,20 +258,34 @@ bool read_bin_goto_object(
   irep_serializationt irepconverter(ic);
   // symbol_serializationt symbolconverter(ic);
 
+  const auto deprecated_binary_version = 5;
+  static_assert(
+    deprecated_binary_version == GOTO_BINARY_VERSION - 1,
+    "The deprecated goto binary version is the one preceding current version");
+
   {
     std::size_t version=irepconverter.read_gb_word(in);
-
-    if(version < GOTO_BINARY_VERSION)
+    bool is_deprecated_version = version == deprecated_binary_version;
+    if(is_deprecated_version)
+    {
+      message.warning()
+        << "The input was compiled with an old version of goto-cc; "
+           "which is deprecated and will be removed in cbmc version 6. "
+           "Please recompile."
+        << messaget::eom;
+    }
+    if(version == GOTO_BINARY_VERSION || is_deprecated_version)
```

Comment:
Is the condition here correct?

It looks like it's still going to try to read in the file if it's a deprecated version, whereas it looks like we're handling the same situation below in the `else if` branch with an error.

Reply:
That is entirely the intended behaviour. The new argument to `read_bin_goto_object` switches between reading the deprecated format and the current format. In this context deprecated means we currently support this, but we are planning to remove the support. Therefore in the case of the deprecated format we warn immediately before reading and then do the appropriate form of reading. In the case of versions older than the deprecated version, the pre-existing error message is reached instead.